### PR TITLE
curvefs/client: adapter lru list for disk cache

### LINKED ISSUE #1075: disk cache(read) use lru list
**Is your feature request related to a problem? (你需要的功能是否与某个问题有关?)**

**Describe the solution you'd like (描述你期望的解决方法)**

**Describe alternatives you've considered (描述你想到的折衷方案)**

**Additional context/screenshots (更多上下文/截图)**

### PR DESCRIPTION
**Pull request by wuhongsong**
### What problem does this PR solve?

Issue Number: close #1075

### CODE REVIEW

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

File: curvefs/src/client/s3/disk_cache_manager.cpp [hunk 1]
```diff
@@ -331,23 +335,23 @@ void DiskCacheManager::TrimCache() {
...
-                cacheWriteFile = cacheWriteFullDir + "/" + *cacheKeyIter;
+                std::string cacheKey;
+                cachedObjName_->GetBack(&cacheKey);
+                if (cacheKey.empty()) {
+                    VLOG(3) << "cachekey is empty";
+                    break;
+                }
+                VLOG(9) << "obj will be removed: " << cacheKey;
+                cacheReadFile = cacheReadFullDir + "/" + cacheKey;
+                cacheWriteFile = cacheWriteFullDir + "/" + cacheKey;
```
Comment (baijiaruo) [inline curvefs/src/client/s3/disk_cache_manager.cpp:354]: If cacheWriteFile exists, why not evict the next？

File: src/common/lru_cache.h [hunk 2]
```diff
@@ -356,6 +356,195 @@ std::shared_ptr<CacheMetrics>
...
+    if (ll_.empty()) {
+        LOG(INFO) << "cache is empty.";
+        return;
+    }
+    *value = ll_.back();
+    return;
+}
+
+template <typename K, typename KeyTraits>
+bool SglLRUCache<K, KeyTraits>::IsCached(const K &key) {
```
Comment (xu-chaojie) [inline src/common/lru_cache.h:464]: if not concern the item evicted, return void is enough.

File: src/common/lru_cache.h [hunk 3]
```diff
@@ -356,6 +356,195 @@ std::shared_ptr<CacheMetrics>
...
+}
+
+template <typename K, typename KeyTraits>
+void SglLRUCache<K, KeyTraits>::Remove(const K &key) {
+    ::curve::common::WriteLockGuard guard(lock_);
+    RemoveLocked(key);
+}
+
+template <typename K, typename KeyTraits>
+bool SglLRUCache<K, KeyTraits>::PutLocked(const K &key) {
```
Comment (xu-chaojie) [inline src/common/lru_cache.h:490]: If you don't care about the evicted item, returning void is enough.

File: src/common/lru_cache.h [hunk 4]
```diff
@@ -356,6 +356,195 @@ std::shared_ptr<CacheMetrics>
...
+ public:
+    /**
+     * @brief Store key to the cache
+     *
+     * @param[in] key
+     *
+     */
+    virtual void Put(const K &key) = 0;
+
+    virtual bool IsCached(const K &key) = 0;
```
Comment (xu-chaojie) [inline src/common/lru_cache.h:370]: comment

File: src/common/lru_cache.h [hunk 5]
```diff
@@ -356,6 +356,195 @@ std::shared_ptr<CacheMetrics>
...
+void SglLRUCache<K, KeyTraits>::MoveToFront(
+    const typename std::list<K>::iterator &elem) {
+    K tmp = *elem;
+    ll_.erase(elem);
+    ll_.push_front(tmp);
+    cache_[tmp] = ll_.begin();
+}
+
+template <typename K, typename KeyTraits>
+bool SglLRUCache<K, KeyTraits>::RemoveOldest() {
```
Comment (xu-chaojie) [inline src/common/lru_cache.h:528]: If you don't care about the evicted item, returning void is enough.

File: src/common/lru_cache.h [hunk 6]
```diff
@@ -356,6 +356,195 @@ std::shared_ptr<CacheMetrics>
...
+    if (ll_.begin() != ll_.end()) {
+        RemoveElement(--ll_.end());
```
Comment (xu-chaojie) [inline src/common/lru_cache.h:530]: add a "return true" in line 333  for me,  there is a bug there

File: curvefs/src/client/s3/disk_cache_manager.cpp [hunk 7]
```diff
@@ -359,28 +363,15 @@ void DiskCacheManager::TrimCache() {
...
-                                << ", file is not exist in cachedObjName"
-                                << ", file is: " << *cacheKeyIter;
-                        cachedObjNameTmp.erase(cacheKeyIter);
-                        continue;
-                    }
-                    cachedObjName_.erase(iter);
+                            << ", file is: " << cacheKey;
+                    break;
                 }
+                cachedObjName_->Remove(cacheKey);
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/disk_cache_manager.cpp:369]: if we remove cachekey first, then come a read, will it cause an concurrency problem?

File: curvefs/src/client/s3/disk_cache_read.cpp [hunk 8]
```diff
@@ -102,7 +102,8 @@ int DiskCacheRead::LinkWriteToRead(const std::string fileName,
     return 0;
 }
 
-int DiskCacheRead::LoadAllCacheReadFile(std::set<std::string> *cachedObj) {
+int DiskCacheRead::LoadAllCacheReadFile(
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/disk_cache_read.cpp:105]: Never find disk read hit the cache!

Comment (xu-chaojie) [inline curvefs/src/client/s3/disk_cache_read.cpp:105]: Never find diskcache read hit the cache!

Comment (wuhongsong) [reply curvefs/src/client/s3/disk_cache_read.cpp:105]: IsCache is been called before Read.

File: curvefs/src/client/s3/disk_cache_manager.cpp (hunk 1 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/disk_cache_manager.cpp:354]: may fix later, the issue is : https://github.com/opencurve/curve/issues/1117

File: src/common/lru_cache.h (hunk 6 above)
Comment (wuhongsong) [reply src/common/lru_cache.h:530]: fix

File: src/common/lru_cache.h (hunk 3 above)
Comment (wuhongsong) [reply src/common/lru_cache.h:490]: fix

File: src/common/lru_cache.h (hunk 5 above)
Comment (wuhongsong) [reply src/common/lru_cache.h:528]: fix

File: curvefs/src/client/s3/disk_cache_manager.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/src/client/s3/disk_cache_manager.cpp:369]: the obj name has unique identification

File: src/common/lru_cache.h (hunk 4 above)
Comment (wuhongsong) [reply src/common/lru_cache.h:370]: fix

**Comment by wuhongsong** [comment]
recheck

**Review by xu-chaojie — APPROVED** [review]

**Review by baijiaruo — APPROVED** [review]

**Pull request merged**